build/lib/src/generate/expected_outputs.dart

```diff
@@ -7,22 +7,115 @@ import '../builder/builder.dart';
 }
 
-Iterable<AssetId> _replaceExtensions(
-        AssetId assetId, String oldExtension, List<String> newExtensions) =>
-    newExtensions.map((n) => _replaceExtension(assetId, oldExtension, n));
-
-AssetId _replaceExtension(
-    AssetId assetId, String oldExtension, String newExtension) {
-  var path = assetId.path;
-  assert(path.endsWith(oldExtension));
-  return AssetId(
-      assetId.package,
-      path.replaceRange(
-          path.length - oldExtension.length, path.length, newExtension));
+// We can safely cache parsed build extensions for each builder since build
+// extensions are required to not change for a builder.
+final _parsedInputs = Expando<List<_ParsedBuildOutput>>();
+
+extension on Builder {
+  List<_ParsedBuildOutput> get parsedExtensions {
+    final existing = _parsedInputs[this];
+    if (existing == null) {
+      return _parsedInputs[this] = [
+        for (final entry in buildExtensions.entries)
+          _ParsedBuildOutput.parse(entry.key, entry.value),
+      ];
+    }
+
+    return existing;
+  }
+}
+
+class _ParsedBuildOutput {
+  static final RegExp _captureGroup = RegExp('{{}}');
+
+  final RegExp _pathMatcher;
+  final List<String> _outputs;
+  final bool _usesCaptureGroup;
+
+  _ParsedBuildOutput(this._pathMatcher, this._outputs, this._usesCaptureGroup);
+
+  factory _ParsedBuildOutput.parse(String input, List<String> outputs) {
+    final groups = _captureGroup.allMatches(input).iterator;
+
+    if (!groups.moveNext()) {
+      // The input does not contain a capture group, so we should simply match
+      // all files whose paths ends with the desired input.
+      return _ParsedBuildOutput(
+          RegExp('${RegExp.escape(input)}\$'), outputs, false);
+    }
+
+    final match = groups.current;
+
+    if (groups.moveNext()) {
+      throw ArgumentError(
+        'The builder declares an input "$input" which contains multiple '
+        'groups ("{{}}"). This is not supported.',
+      );
+    }
+
+    // When using a capture group in the build input, it must also be used in
+    // every output to ensure outputs have unique names.
+    for (final output in outputs) {
+      if (!_captureGroup.hasMatch(output)) {
+        throw ArgumentError(
+          'A builder declares an input "$input" using a capture group. '
+          'It is required that all of its output also refer to that capture '
+          'group. However, "$output" does not.',
+        );
+      }
+    }
+
+    final regexBuffer = StringBuffer();
+    // Start writing the regex up to the capture group
+    if (input.startsWith('^')) {
+      // Build inputs starting with `^` start matching at the beginning of the
+      // input path.
+      regexBuffer
+        ..write('^')
+        ..write(RegExp.escape(input.substring(1, match.start)));
+    } else {
+      regexBuffer.write(RegExp.escape(input.substring(0, match.start)));
+    }
+
+    regexBuffer
+      ..write('(.+)') // capture group
```

Comment:
A bit of an open question if this should be `.+` or `.*` I suppose, but I think probably requiring some matching text here is the correct choice.